pkg/flowaggregator/s3uploader/s3uploader.go

```diff
@@ -0,0 +1,464 @@
+		if err != nil {
+			return fmt.Errorf("error when loading AWS config: %w", err)
+		}
+		p.region = region
+		p.awsS3Client = s3.NewFromConfig(cfg)
+		p.awsS3Uploader = s3manager.NewUploader(p.awsS3Client)
+	}
+	return nil
+}
+
+func (p *S3UploadProcess) SetUploadInterval(uploadInterval time.Duration) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	p.uploadInterval = uploadInterval
+	if p.uploadTicker != nil {
+		p.uploadTicker.Reset(p.uploadInterval)
+	}
+}
+
+func (p *S3UploadProcess) CacheRecord(record ipfixentities.Record) {
+	r := flowrecord.GetFlowRecord(record)
+	p.bufferMutex.Lock()
+	defer p.bufferMutex.Unlock()
+	// If the number of pending records in the buffer reaches bufferSize,
+	// add the buffer to bufferQueue.
+	if int32(p.cachedRecord) >= p.maxRecordPerFile {
+		p.appendFileToQueue()
+	}
+	p.writeFile(r)
+	p.cachedRecord += 1
+}
+
+func (p *S3UploadProcess) Start() {
+	p.startExportProcess()
+}
+
+func (p *S3UploadProcess) Stop() {
+	p.stopExportProcess(true)
+}
+
+func (p *S3UploadProcess) startExportProcess() {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	if p.exportProcessRunning {
+		return
+	}
+	p.exportProcessRunning = true
+	p.uploadTicker = time.NewTicker(p.uploadInterval)
+	p.stopCh = make(chan stopPayload, 1)
+	p.exportWg.Add(1)
+	go func() {
+		defer p.exportWg.Done()
+		p.flowRecordPeriodicCommit()
+	}()
+}
+
+func (p *S3UploadProcess) stopExportProcess(flushQueue bool) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	if !p.exportProcessRunning {
+		return
+	}
+	p.exportProcessRunning = false
+	defer p.uploadTicker.Stop()
+	p.stopCh <- stopPayload{
+		flushQueue: flushQueue,
+	}
+	p.exportWg.Wait()
+}
+
+func (p *S3UploadProcess) flowRecordPeriodicCommit() {
+	klog.InfoS("Starting S3 exporting process")
+	ctx := context.Background()
+	for {
+		select {
+		case stop := <-p.stopCh:
+			klog.InfoS("Stopping S3 exporting process")
+			if !stop.flushQueue {
+				return
+			}
+			ctx, cancelFn := context.WithTimeout(ctx, bufferFlushTimeout)
+			defer cancelFn()
+			err := p.batchUploadAll(ctx)
+			if err != nil {
+				klog.ErrorS(err, "Error when doing batchUploadAll on stop")
+			}
+			return
+		case <-p.uploadTicker.C:
+			err := p.batchUploadAll(ctx)
+			if err != nil {
+				klog.ErrorS(err, "Error when doing batchUploadAll on triggered timer")
+			}
+		}
+	}
+}
+
+// batchUploadAll uploads all CSV files cached in bufferQueue. Returns error if
+// encountered. Cached files will be removed after a successful upload. Otherwise,
+// they will be kept in the bufferQueue.
+func (p *S3UploadProcess) batchUploadAll(ctx context.Context) error {
+	p.bufferMutex.Lock()
+	if p.cachedRecord != 0 {
+		p.appendFileToQueue()
+	}
+	p.bufferMutex.Unlock()
+
+	p.queueMutex.Lock()
+	if len(p.bufferQueue) == 0 {
+		p.queueMutex.Unlock()
+		return nil
+	}
+	defer p.queueMutex.Unlock()
+	uploaded := 0
+	for _, buf := range p.bufferQueue {
+		bufCopy := bytes.NewBuffer(buf.Bytes())
```

Comment:
the whole point of recent changes is to avoid copies, so why do we copy the buffer?

Reply:
Even if the upload is failed, the buffer will be cleared. So I was making the copy to help restore the data in the failed buffer. 

However, I think we can achieve the similar effect by making a copy of the queue, as you suggested below - `buffersToUpload `.